plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/ClosedProjectQueue.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) 2018 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ui.building;
+
+import java.util.Deque;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import org.eclipse.xtext.builder.impl.ToBeBuilt;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Singleton;
+
+/**
+ * A queue of projects that are supposed to be removed from the Xtext index.
+ */
+@SuppressWarnings("restriction")
+@Singleton
+public class ClosedProjectQueue {
+
+	/**
+	 * Encapsulates the changes that need to be performed after one or more projects have been removed / closed.
+	 */
+	class Task {
+		/**
+		 * The names of the projects as encapsulated by this task.
+		 */
+		public final Set<String> projectNames;
+		/**
+		 * The built data for this task.
+		 */
+		public final ToBeBuilt toBeBuilt;
+
+		private Task(Set<String> projectNames, ToBeBuilt toBeBuilt) {
+			this.projectNames = projectNames;
+			this.toBeBuilt = toBeBuilt;
+		}
+
+		/**
+		 * Returns true if the is an empty task.
+		 *
+		 * @return true if empty.
+		 */
+		public boolean isEmpty() {
```

Comment:
since we are exposing `toBeBuilt` we can skip this method.  it is used only in `CloseProjectTaskScheduler.processClosedProjects()` which is accessing `toBeBuilt` directly anyway.